### ISSUE TITLE
Feature Request: `cols_label_with()` relabeling by function

### ISSUE THREAD

**Issue by christopherkenny** [rstudio/gt#1151]
## Prework

- [x] Read and abide by **gt**'s [code of conduct](https://www.contributor-covenant.org/version/2/0/code_of_conduct/) and [contributing guidelines](https://github.com/rstudio/gt/blob/master/.github/CONTRIBUTING.md).
- [x] Search for duplicates among the [existing issues](https://github.com/rstudio/gt/issues) (both open and closed).

## Proposal

Implement a `cols_label_with()` that parallels `rename_with()`'s relationship with `rename()`. Ideally, it would  operate on the _existing_ labels, rather than names directly so that
  - you can iteratively build up with multiple functions
  - `tidyselect`ions can overlap without undoing prior work
  - weird transformations can be done manually but still wrapped in `md` or `html` via a later function call 

## Related Issues

- #701: Alternate to allowing `tidyselect`ion on the right hand side 
  - makes implementation easier and arguably easier to use, though does mean a second function
  - helps resolve the potential confusion when a `tidyselect`ion captures a column more than once.
- #817: Alternate to adding a `.fn` argument to `cols_label()`
- #1146: Would allow you to use `cols_label_with(fn = md)` to avoid places where the `from_markdown` class is lost

## Proof of Concept

```
#' Relabel columns with a function
#'
#' @param .data A table object that is created using the [gt()] function.
#' @param fn The function to apply to existing column labels.
#' @param columns The column names to apply `fn` to. Applies to all columns by default.
#' @param ... arguments to pass to
#'
#' @return An object of class `gt_tbl`.
#' @export
#'
#' @section Examples:
#'
#' Use [`countrypops`] to create a **gt** table. Relabel all the table's columns
#' with the `cols_label_with()` function to improve its presentation. It applies
#' to existing labels, rather than just names, so we can make more difficult
#' changes directly.
#'
#' ```r
#' countrypops %>%
#'   dplyr::select(-contains("code")) %>%
#'   dplyr::filter(country_name == "Mongolia") %>%
#'   tail(5) %>%
#'   gt() %>%
#'   cols_label(
#'     country_name = "Name"
#'   ) %>%
#'   cols_label_with(
#'     tools::toTitleCase
#'   )
#' ```
cols_label_with <- function(.data, fn, columns = everything(), ...) {
  fn <- rlang::as_function(fn)
  cols <- tidyselect::eval_select(
    rlang::enquo(columns),
    dt_data_get(data = .data),
    allow_rename = FALSE
  )

  names <- dt_boxhead_get_vars(data = .data)
  old_label_list <-  dt_boxhead_get(data = .data)$column_label[cols]

  nm_labels_list <- names[cols]
  labels_list <- lapply(old_label_list, function(x) fn(x, ...))

  if (!all(vapply(labels_list, is_character, logical(1L)))) {
    cli::cli_abort(
      "{.arg fn} must return a character vector, not {.obj_type_friendly {labels_list}}."
    )
  }
  if (length(labels_list) != length(nm_labels_list)) {
    cli::cli_abort(
      "{.arg fn} must return a vector of length {length(nm_labels_list)}, not {length(labels_list)}."
    )
  }

  # If no labels remain after filtering, return the data
  if (length(nm_labels_list) == 0) {
    return(.data)
  }

  for (i in seq_along(labels_list)) {
    .data <-
      dt_boxhead_edit_column_label(
        data = .data,
        var = nm_labels_list[i],
        column_label = labels_list[[i]]
      )
  }

  .data

}
```

Example of how this would work:

``` 
countrypops %>%
  dplyr::select(-contains("code")) %>%
  dplyr::filter(country_name == "Mongolia") %>%
  tail(5) %>%
  gt() %>%
  cols_label(
    country_name = "Name"
  ) %>%
  cols_label_with(
    tools::toTitleCase
  )
``` 

Output:
![image](https://user-images.githubusercontent.com/28026893/204964284-bc2cb8c2-cd0a-4267-a0e5-35ca75bbd56c.png)

Or with a bit more complication, `~`, `...` and `tidyselect` (a bit contrived, but simple for quick experimentation)

```
countrypops %>%
  dplyr::filter(country_name == "Mongolia") %>%
  tail(5) %>%
  gt() %>%
  cols_label_with(
    fn = stringr::str_replace,
    pattern = '_',
    replacement = '\n'
  ) %>%
  cols_label_with(
    ~tools::toTitleCase(.) %>% md()
  ) %>% 
  cols_label_with(
    stringr::str_remove,
    pattern = '_\\d+',
    columns = -contains('3')
  )
```

![image](https://user-images.githubusercontent.com/28026893/204965326-8377fd99-777e-4430-9d21-e0202bd78e27.png)

This is a bit of an opinionated approach to resolving these types of issues, particularly in that it builds off of existing _labels_ rather than column names, as that what makes sense to me conceptually. Happy to discuss and/or propose a PR if it this type of approach is useful.

**Comment by rich-iannone** [comment]
Okay, before I say anything else, best Issue ever! I’m just really impressed with the level of thought and care you put into this. Totally agree with the need and the approach taken here. Thank you also for the suggested implementation!

Seriously, if you wanted to take the last step and submit a PR, I could help you take it over the finish line with anything that’s needed.

**Comment by christopherkenny** [comment]
Thank you! I've added #1158, which implements the feature and adds some tests (though I expect more tests may be helpful).

**Comment by rich-iannone** [comment]
Fixed by #1158. Thanks again for all of this!

**Comment by christopherkenny** [comment]
Happy to help. Thanks for the quick turnaround on review and approval! This feature is super useful for election data tables, so I'm very happy to have it integrated.
